[MM-51412] - In-product banner to communicate Cloud Free deprecation

AGMETEOR opened the pull request:
#### Summary
Adds banners that communicate to admin about cloud free deprecation

#### Screenshots
<img width="1792" alt="Screenshot 2023-03-28 at 15 40 11" src="https://user-images.githubusercontent.com/28563179/228239973-6921fb20-0dc9-4bb8-b6db-10b96c55c5b9.png">

<img width="1792" alt="Screenshot 2023-03-28 at 15 44 27" src="https://user-images.githubusercontent.com/28563179/228240031-5587df5d-0944-426f-abd8-a4995c8d3316.png">

<img width="1792" alt="Screenshot 2023-03-28 at 16 09 55" src="https://user-images.githubusercontent.com/28563179/228246995-a10dbc2a-7071-4149-ab34-385943f24c70.png">


<img width="1792" alt="Screenshot 2023-03-28 at 15 40 25" src="https://user-images.githubusercontent.com/28563179/228240013-bca9015e-433a-4a64-9df2-964f0f0a4715.png">


#### Ticket Link
https://mattermost.atlassian.net/browse/MM-51412

#### Release Note

```release-note
none
```

AGMETEOR commented:
/update-branch

mattermost-build commented:
E2E tests not automatically triggered, because PR has no approval yet. Please ask a developer to review and then try again to attach the QA label.

AGMETEOR commented on webapp/channels/src/components/admin_console/billing/billing_subscriptions/to_paid_plan_nudge_banner.tsx:
```diff
@@ -0,0 +1,237 @@
+    Zero = '0'
+}
+
+const cloudFreeCloseMoment = '20230428'; // TBD, final day of cloud free
```
To test banners, update this date accordingly

AGMETEOR commented:
Split available here https://app.split.io/org/e23fa240-f3b6-11ea-b78e-12a8e42e6081/ws/e244d260-f3b6-11ea-b78e-12a8e42e6081/splits/09ad5ad0-cd7f-11ed-b0c0-7abdf24ac0eb/env/e24b3b00-f3b6-11ea-b78e-12a8e42e6081/definition

ConorMacpherson commented on webapp/channels/src/components/admin_console/billing/billing_subscriptions/to_paid_plan_nudge_banner.tsx:
```diff
@@ -0,0 +1,242 @@
+    const enabled = useSelector(cloudFreeDeprecationUIEnabled);
+    const currentProductStarter = product?.sku === CloudProducts.STARTER;
+
+    const now = moment(Date.now());
```
Nit:

Very minor and likely not very noticeable, but we could instead do:

```typescript
const now = moment().startOf('day')
```

Since the current implementation will include the time (rather than just the date), the time will be a part of the diff. 

So, if the date string that we have set here for cloud to be deprecated was `2023-03-30` with today being `2023-03-29`, the `daysToCloudFreeEnd` would actually appear to be 0, even if there is some time left in the day until deprecation.

It might cause some confusion when the banner message states `Cloud Starter will be deprecated in 0 days. Upgrade to a paid plan or contact sales.`

This might also require a change here: https://github.com/mattermost/mattermost-server/pull/22622/files#diff-cf8060a56e72d942b4fe0feb8925d56ebe7430e89817c2510c4aa0d4aa3222b5R154

To check if `daysToCloudFreeEnd == 0` rather than `daysToCloudFreeEnd < 0` so we display a message stating that the plan was deprecated once we hit 0 days.

ConorMacpherson reviewed (dismissed):
Nice work! Approving so I don't remain as a blocker when Nathaniel's suggested changes re made.

yasserfaraazkhan commented:
/update-branch

AGMETEOR commented:
/update-branch

neallred commented on webapp/channels/src/components/admin_console/billing/billing_subscriptions/to_paid_plan_nudge_banner.tsx:
```diff
@@ -0,0 +1,242 @@
+    };
+
+    useEffect(() => {
+        if (!snoozeInfo.show) {
```
Ahh yep! Thanks for the clarification.

neallred commented on webapp/channels/src/components/admin_console/billing/billing_subscriptions/to_paid_plan_nudge_banner.test.tsx:
```diff
@@ -0,0 +1,239 @@
+
+        renderWithIntlAndStore(<ToPaidPlanBannerDismissable/>, state);
+
+        expect(screen.getByTestId('cloud-free-deprecation-announcement-bar')).toBeInTheDocument();
```
Nit: all the `getBy*` selectors (like `getByTestId` ) are assertions all by themselves, so we can omit the wrapping `expect(...).toBeInTheDocument()`. The test will fail if it is not found without that (as seen by the example below that expects it to throw in the not found case)

neallred commented on webapp/channels/src/components/admin_console/billing/billing_subscriptions/to_paid_plan_nudge_banner.test.tsx:
```diff
@@ -0,0 +1,239 @@
+
+        renderWithIntlAndStore(<ToPaidNudgeBanner/>, state);
+
+        expect(screen.getByTestId('cloud-free-deprecation-alert-banner')).toBeInTheDocument();
```
Ditto can omit `expect(...).toBeInTheDocument()`

neallred reviewed (approved):
LGTM, thanks @AGMETEOR !

yasserfaraazkhan reviewed (approved):
Thank you Allan
have tested this on spin wick and locally. Its working as expected.

AGMETEOR commented on webapp/channels/src/components/admin_console/billing/billing_subscriptions/to_paid_plan_nudge_banner.tsx:
```diff
@@ -0,0 +1,246 @@
+    Zero = '0'
+}
+
+const cloudFreeCloseMoment = '20230327'; // TBD, final day of cloud free
```
Here.

AGMETEOR commented:
/update-branch

AGMETEOR commented:
Hey @ConorMacpherson and @neallred , reviews went stale

yasserfaraazkhan commented on webapp/channels/src/components/admin_console/billing/billing_subscriptions/to_paid_plan_nudge_banner.tsx:
```diff
@@ -0,0 +1,246 @@
+    const description = (
+        <FormattedMessage
+            id='cloud_billing.nudge_to_paid.description'
+            defaultMessage='Cloud Starter will be deprecated in {days} days. Upgrade to a paid plan or contact sales.'
```
should it be `Cloud Free`
